tpmutil/mssim/mssim.go

```diff
@@ -0,0 +1,227 @@
+	tpmSendCommand    = 8
+	tpmSignalNVOn     = 11
+	tpmSessionEnd     = 20
+)
+
+// Config holds configuration parameters for connecting to the simluator.
+type Config struct {
+	// Addresses of the command and platform handlers.
+	//
+	// Defaults to port 2321 and 2322.
+	CommandAddress  string
+	PlatformAddress string
+
+	// DialContext is used to connect to the simluator ports.
+	//
+	// If nil, a reasonable default is provided.
+	DialContext func(ctx context.Context, network, address string) (net.Conn, error)
+}
+
+// Open attempts to initialize the simulator, power cycling the platform.
+func Open(ctx context.Context, config Config) (*Conn, error) {
+	cmdAddr := config.CommandAddress
+	if cmdAddr == "" {
+		cmdAddr = "127.0.0.1:2321"
+	}
+
+	platformAddr := config.PlatformAddress
+	if platformAddr == "" {
+		platformAddr = "127.0.0.1:2322"
+	}
+
+	dial := config.DialContext
+	if dial == nil {
+		dial = (&net.Dialer{}).DialContext
+	}
+
+	conn, err := dial(ctx, "tcp", platformAddr)
+	if err != nil {
+		return nil, fmt.Errorf("dial platform address: %v", err)
+	}
+	defer conn.Close()
+
+	// Startup the simulator.
+
+	buff := [4]byte{}
+	if err := sendPlatformCommand(conn, buff, tpmSignalPowerOff); err != nil {
+		return nil, fmt.Errorf("power off platform command failed: %v", err)
+	}
+	if err := sendPlatformCommand(conn, buff, tpmSignalPowerOn); err != nil {
+		return nil, fmt.Errorf("power on platform command failed: %v", err)
+	}
+	if err := sendPlatformCommand(conn, buff, tpmSignalNVOn); err != nil {
+		return nil, fmt.Errorf("nv on platform command failed: %v", err)
+	}
+
+	binary.BigEndian.PutUint32(buff[:], tpmSessionEnd)
+	if _, err := conn.Write(buff[:]); err != nil {
+		return nil, fmt.Errorf("shutdown platform connection failed: %v", err)
+	}
+
+	return &Conn{addr: cmdAddr, dial: dial}, nil
+}
+
+// sendPlatformCommand sends device management commands to the simulator.
+//
+// See: "D.4.3.2. PlatformServer()"
+func sendPlatformCommand(conn net.Conn, buff [4]byte, u uint32) error {
+	binary.BigEndian.PutUint32(buff[:], u)
+	if _, err := conn.Write(buff[:]); err != nil {
+		return fmt.Errorf("write platform command: %v", err)
+	}
+	if _, err := io.ReadFull(conn, buff[:]); err != nil {
+		return fmt.Errorf("read platform command: %v", err)
+	}
+	rc := binary.BigEndian.Uint32(buff[:])
+	if rc != 0 {
+		return fmt.Errorf("unexpected platform command response code: 0x%x", rc)
+	}
+	return nil
+}
+
+const (
+	connStateIdle    = 0
+	connStateRunning = 1
+)
+
+// Conn is a Microsoft Simulator client that can be used as a connection for the
+// tpm2 package.
+//
+//		client, err := mssim.Open(mssim.Config{})
+//		if err != nil {
+//			// handle error
+//		}
+//		defer client.Close()
+//
+//		randBytes, err := tpm2.GetRandom(client, 8)
+//
+type Conn struct {
+	addr string
+
+	dial func(ctx context.Context, network, address string) (net.Conn, error)
+
+	// Track if there's an outgoing connection to detect concurrent use.
+	connState uint32
```

Comment:
Dial `conn` in `Open` and get rid of this.
All other TPM connection implementations don't support concurrency either.

Reply:
done